_posts/2017-01-30-docker-frontend-tools.md

```diff
@@ -0,0 +1,146 @@
+
+One piece of advice I have for building images for local development is that while Alpine Linux based images may be much smaller in size, the bare-bones nature and lack of common packages makes it much more difficult to build upon. For that reason, we based our image on the standard `DebianJessie` Node image rather than the Alpine variant.
+
+Now a Docker purist might critique this image and recommend only “one process per container”. However, since this image is for local development, isn't being used to deploy a production app, and encapsulates all of the applications required by the Pattern Lab Starter theme, we felt comfortable with this approach.
+
+### Using the image
+
+To use this image, we specified it in our project's `docker-compose.yml` file by adding the following lines to the services section:
+
+```yaml
+node-php-composer:
+ image: savaslabs/node-php-composer:1.2
+ ports:
+   - "3050:3050"
+   - "3001:3001"
+ volumes_from:
+   - php
+```
+
+This defines the configuration that is applied to a `node-php-composer` container when spun up. This configuration:
+
+- Specifies that the container should be created from the `savaslabs/node-php-composer` image that we built and referenced previously
+- Maps the container ports to our host ports so that we can access the Pattern Labs style guide locally
+- Mounts the project files (that are mounted to the php container) so that they are accessible to the container.
+
+With this service defined in the `docker-compose.yml` we can start using the theme!
+
+First we spin up the Docker containers by running `docker-compose up -d`.
+
+Once the containers are running, we shell directly into the theme directory of the `node-php-composer` container by running the command:
+
+```bash
+docker-compose run --rm --service-ports -w /var/www/html/web/themes/custom/pattern_lab_starter node-php-composer /bin/bash
+```
+
+We use the `--service-ports` option [to ensure the ports used for serving the style guide are mapped to the host](https://github.com/docker/compose/issues/1259#issuecomment-90878095).
+
+Once inside the container in the theme directory, we install the theme's dependencies and serve the style guide by running the following commands:
+
+```bash
+npm install --unsafe-perm
+npm start
+```
+Voila! Once `npm start` is running we can access the Pattern Lab style guide at the URL's that are outputted, for example `http://localhost:3050/pattern-lab/public/`.
+
+Note: Docker runs containers as root, so we use the `--unsafe-perm` flag to run `npm install` with root privileges. This is okay for local development, but would be a security risk if deploying the container to production. For information on running the container as an unprivileged user, [see this documentation](https://github.com/nodejs/docker-node/blob/master/docs/BestPractices.md#non-root-user).
```

Comment:
Aah, thanks for explaining @dmurphy1 @kostajh.

I realize I misunderstood the point of Docker -- I thought we were containerizing all applications and running those same containers locally _and in production_ to have identical environments. I see now you're saying our local Docker containers just capture the version of applications on the production servers, and together mimic the production environment as a whole. 

So, the security risk only applies to the docker containers in a local setting because they don't exist on production!